crates/katana/executor/src/blockifier/utils.rs

```diff
@@ -87,6 +99,103 @@ pub fn estimate_fee(
         .collect::<Result<Vec<_>, _>>()
 }
 
+/// Simulate a transaction's execution on the state
+pub fn simulate_transaction(
+    transaction: ExecutableTxWithHash,
+    block_context: &BlockContext,
+    state: Box<dyn StateProvider>,
+    validate: bool,
+    charge_fee: bool,
+) -> Result<SimulatedTransaction, TransactionExecutionError> {
+    let state = CachedStateWrapper::new(StateRefDb::from(state));
+    let result = TransactionExecutor::new(
+        &state,
+        block_context,
+        charge_fee,
+        validate,
+        vec![transaction.clone()].into_iter(),
```

Comment:
This whole function should just accept a `Vec<ExecutableTxWithHash>` and returns the results of all the simulated txs, considering the executor here already accepts an iterator. Then we can zip the output of the executor with the transactions, and map it to perform the rest of the function.

Reply:
From your comment I feel like you want me to pass the `Vec<Executables>` to the executor. I'm not entirely sure that is correct. What I am not sure about is: does `simulate_transactions` simulate ALL the `n` transactions one after the other on the state `S_i` from the initial  state `S_0`, leading to `S_n` or do you simulate each transaction separately on state `S_0`, leading to states `S_1`, `S_1'`, `S_1''`,...?

Reply:
Let me know, I took the second option

Reply:
this is what the `simulateTransactions` [spec](https://github.com/starkware-libs/starknet-specs/blob/49665932a97f8fdef7ac5869755d2858c5e3a687/api/starknet_trace_api_openrpc.json#L58) says:

> a sequence of transactions to simulate, running each transaction on the state resulting from applying all the previous ones

so i believe this is referring to the first option.

Reply:
Completely missed that, thanks!